mhlib/app.go

```diff
@@ -269,3 +293,81 @@ func (a *App) render(configFile string) (*string, *string, *[]byte, error) {
 
 	return &chart, &chartVersion, &overrides, nil
 }
+
+
+func selfRender(templateValuesStr string, enableGomplate bool) (string, error) {
+	/*
+	This function will accept an input string and run it through the
+	  templating engine as both the values dictionary as well as the
+	  template string.  It will repeat the process until the templating
+	  result stops changing.
+	*/
+
+	type Values map[string]interface{}
+
+	lastRender := templateValuesStr
+	for i := 0; i < 10; i++ {
+
+		// Unmarshal the file as a values dict
+		vals := Values{}
+		err := yaml.Unmarshal([]byte(templateValuesStr), &vals)
+		if err != nil { panic(err) }
+
+		tmpl := template.New("SelfTemplate")
+		tmpl.Delims("[[", "]]")
+
+		if enableGomplate {
+			// Read the defined datasources and datasourcehaders
+			//   from the values dict
+			dataSources := []string{}
+			dataSourceHeaders := []string{}
+			if _, ok := vals["gomplate"]; ok {
+				tmp := vals["gomplate"].(map[string]interface{})
+				if _, ok := tmp["datasources"]; ok {
+					dataSources = convertInterfaceListToStringList(
+						tmp["datasources"].([]interface{}))
+				}
+				if _, ok := tmp["datasources"]; ok {
+					dataSourceHeaders = convertInterfaceListToStringList(
+						tmp["datasourceheaders"].([]interface{}))
+				}
+
+			}
+
+			// Access gomplate datasources and function library
+			d, err := data.NewData(dataSources, dataSourceHeaders)
+			if err != nil { panic(err) }
+
+			// Configure go/text/template to use gomplate
+			//   datasources and function library.
+			tmpl.Option("missingkey=error")
+			tmpl.Funcs(gomplate.Funcs(d))
+		}
+
+		// Run the the file through the tempating engine as both values
+		//   file and template file
+		tmpl.Parse(string(templateValuesStr))
+		if err != nil { panic(err) }
+		out := new(bytes.Buffer)
+		err = tmpl.Execute(out, vals)
+		if err != nil { panic(err) }
+
+		newRender := out.String()
+		if lastRender == newRender {
+			return newRender, nil // self-templating succeeded
+		} else {
+			lastRender = newRender
+			templateValuesStr = newRender
+		}
+	}
+
+	return templateValuesStr, errors.New("Self-templating failed")
+}
+
+func convertInterfaceListToStringList(l []interface{}) ([]string) {
```

Comment:
Should this function include an error return?

Reply:
converInterfaceListToStringList uses `fmt.sprint` which will always succeed in printing the string representation of any object.

Reply:
Function deleted, with @drekle 's feedback to use Unmarshal with typed target struct